src/raw/mod.rs

```diff
@@ -1055,224 +889,566 @@ impl<T, A: Allocator + Clone> RawTable<T, A> {
+        capacity: usize,
+        fallibility: Fallibility,
+    ) -> Result<Self, TryReserveError> {
+        if capacity == 0 {
+            Ok(Self::new_in(alloc))
+        } else {
+            unsafe {
+                let buckets =
+                    capacity_to_buckets(capacity).ok_or_else(|| fallibility.capacity_overflow())?;
+
+                let result = Self::new_uninitialized(alloc, table_layout, buckets, fallibility)?;
+                result.ctrl(0).write_bytes(EMPTY, result.num_ctrl_bytes());
+
+                Ok(result)
+            }
+        }
+    }
+
+    /// Searches for an empty or deleted bucket which is suitable for inserting
+    /// a new element and sets the hash for that slot.
+    ///
+    /// There must be at least 1 empty bucket in the table.
+    #[inline]
+    unsafe fn prepare_insert_slot(&self, hash: u64) -> (usize, u8) {
+        let index = self.find_insert_slot(hash);
+        let old_ctrl = *self.ctrl(index);
+        self.set_ctrl_h2(index, hash);
+        (index, old_ctrl)
+    }
+
+    /// Searches for an empty or deleted bucket which is suitable for inserting
+    /// a new element.
+    ///
+    /// There must be at least 1 empty bucket in the table.
+    #[inline]
+    fn find_insert_slot(&self, hash: u64) -> usize {
+        let mut probe_seq = self.probe_seq(hash);
+        loop {
+            unsafe {
+                let group = Group::load(self.ctrl(probe_seq.pos));
+                if let Some(bit) = group.match_empty_or_deleted().lowest_set_bit() {
+                    let result = (probe_seq.pos + bit) & self.bucket_mask;
+
+                    // In tables smaller than the group width, trailing control
+                    // bytes outside the range of the table are filled with
+                    // EMPTY entries. These will unfortunately trigger a
+                    // match, but once masked may point to a full bucket that
+                    // is already occupied. We detect this situation here and
+                    // perform a second scan starting at the begining of the
+                    // table. This second scan is guaranteed to find an empty
+                    // slot (due to the load factor) before hitting the trailing
+                    // control bytes (containing EMPTY).
+                    if unlikely(is_full(*self.ctrl(result))) {
+                        debug_assert!(self.bucket_mask < Group::WIDTH);
+                        debug_assert_ne!(probe_seq.pos, 0);
+                        return Group::load_aligned(self.ctrl(0))
+                            .match_empty_or_deleted()
+                            .lowest_set_bit_nonzero();
+                    }
+
+                    return result;
+                }
+            }
+            probe_seq.move_next(self.bucket_mask);
+        }
+    }
+
+    #[allow(clippy::mut_mut)]
+    unsafe fn prepare_rehash_in_place<'s>(
+        &'s mut self,
+        needs_drop: bool,
+        drop: fn(&mut Self, usize),
```

Comment:
This looks very suspicious to me. Are you sure this is actually necessary to improve compile times?

I would have expected the guard to be optimized away most of the time since `Hash::hash` almost never panics (and the compiler knows this since it inlines the hash function).

Also, this function is non-generic so the lack of a `#[inline]` makes it impossible to inline.

Reply:
> This looks very suspicious to me. Are you sure this is actually necessary to improve compile times?

Can't say for sure since `llvm-lines` only indirectly measures compile times, in the llvm-lines I do not see if the guard itself would get optimized away.  However I use a `fn` argument here since the `drop` function would practically never be called since the non-panic path will always `forget` the guard.

Reply:
However with this you are forcing LLVM to perform full codegen on the drop function, whereas it could previously have been eliminated in one of the early LLVM passes as unreachable code.

Could you try moving the guard creation back into `rehash_in_place` to see how much it affects performance?

Reply:
Moving the guard out forces the guard to be specialized for each `T` and increases the number of llvm lines by ~1% . Wall time is to small and has to much variance to be measured, possibly a rust perf run could show something but probably not.

I suppose llvm may codegen the small drop closure eagerly but I think that on balance, having the rest of the `drop` closure in the drop guard being only instantiated once per allocator to outweigh that (my test only has 4 different hashmaps, the wins on grow as you have more hashmaps).

```
               |self_: &mut RawTableInner<A>, index| {
                    self_.bucket::<T>(index).drop();
                },
```

Reply:
The `drop` closure still needs to be instantiated for each `T`. The only thing we are saving is that `guard` and `Scopeguard::drop` are only instantiated for each `A` instead of for each `T`.

Since `Scopeguard::drop` is likely to be optimized away as unreachable, I expect this change to have negligible effect on compile times while improving the performance of `rehash_in_place`.

Could you try a perf run with the guard moved back into `rehash_in_place`? If the result is worse than before then I'm happy to merge this PR as it is.

Reply:
## guard directly in rehash_in_place

https://perf.rust-lang.org/compare.html?start=84864bfea9c00fb90a1fa6e3af1d8ad52ce8f9ec&end=3ca26a5875f0c4d0522408aa4284b042697b23e4

## guard in prepare_rehash_in_place

https://perf.rust-lang.org/compare.html?start=25f39fe80293f77bd86f64a1261a3e2c0ca23847&end=97c7b0501b746bb7191e480b2076cf601f6d0ca4

Seems to be a bit better on average to keep the guard directly in `rehash_in_place` (even if the biggest improvement is a bit smaller).

Reply:
Considering `rehash_in_place` isn't tested I'd be inclined to remove it entirely.